File: javascript/anagram/anagram.js
const arrange = string => [...string.toLowerCase()].sort().join();

export class Anagram {
    constructor(word) {
        this.word = word;
    }

    matches(candidates) {
        return candidates
            .filter(this.isAnagram.bind(this));
    }

    isAnagram(candidate) {
        return arrange(candidate) === arrange(this.word)
            && candidate.toLowerCase() !== this.word.toLowerCase();
    }
}

File: java/anagram/src/main/java/Anagram.java
import java.util.List;
import java.util.ArrayList;
import java.util.stream.Collectors;

public class Anagram {
    private String word, wordArranged;

    Anagram(String word) {
        this.word = word;
        this.wordArranged = arrange(word);
    }

    List<String> match(List<String> candidates) {
        return candidates
            .stream()
            .filter(this::isAnagram)
            .collect(Collectors.toList());
    }

    private boolean isAnagram(String candidate) {
        return arrange(candidate).equals(wordArranged)
            && !candidate.equalsIgnoreCase(word);
    }

    private static String arrange(String word) {
        return word
            .toLowerCase()
            .chars()
            .sorted()
            .collect(StringBuilder::new,
                     StringBuilder::appendCodePoint,
                     StringBuilder::append)
            .toString();
    }
}
